Add support for HTTP and WebSocket API Types

TehNrd opened the pull request:
<!--- Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests --->
Fixes #312 #207 #223  

**Description of Issue Fixed**
This PR adds support for both HTTP and Websocket API Types for the AWS API Gateway. For these types, domain names must use the ApiGatewayV2 AWS API.

This PR essentially bulkifies the entire process of defining and creating domains. Instead of config values being defined on variables in the index.ts file an array of DomainConfigs is created. Then for all operations it will process all of the domain configs. This allows the mixing of different API types and domains in one serverless.yml file. 

It is backward compatible with the current customdomain configuration format. It is also structured in a way that will easily scale should AWS add new API types in the future.

**This can be tested by adding the following branch to your dependencies.**
```
"dependencies": {
  "serverless-domain-manager": "tehnrd/serverless-domain-manager#http-ws-support-BETA"
},
```

**Changes proposed in this pull request**:

* Static global vars have been moved to Globals.ts to allow easy access across multiple files
* Domain configuration options have been removed from index.ts. A new class DomainConfig.ts is created from the plugin configuration. These are placed into an array of domains in index.ts.
* Methods that process many domain configurations at once like `createDomains()` have been bulkified to process all DomainConfigs in the `domains` array.
* Methods that process one domain at a time like `createCustomDomain()` now take an argument of a `DomainConfig`.
* The configuration is backward compatible but still supports new API types with apiType option. This works well for projects that have a single service per serverless.yml file
   ```yaml
    custom:
        customDomain:
            domainName: serverless.foo.com
            stage: ci
            basePath: api
            certificateName: '*.foo.com'
            createRoute53Record: true
            endpointType: 'regional'
            securityPolicy: tls_1_2
            apiType: rest
    ```
* A new configuration format is also added that allows defining multiple API types along with corrresponding domain configurations.
    ```yaml
    custom:
        customDomain:
            rest:
                domainName: rest.serverless.foo.com
                stage: ci
                basePath: api
                certificateName: '*.foo.com'
                createRoute53Record: true
                endpointType: 'regional'
                securityPolicy: tls_1_2
            http:
                domainName: http.serverless.foo.com
                stage: ci
                basePath: api
                certificateName: '*.foo.com'
                createRoute53Record: true
                endpointType: 'regional'
                securityPolicy: tls_1_2
            websocket:
                domainName: ws.serverless.foo.com
                stage: ci
                basePath: api
                certificateName: '*.foo.com'
                createRoute53Record: true
                endpointType: 'regional'
                securityPolicy: tls_1_2
    ```

Unit tests have been updated and a few new ones have been added. I also performed some functional testing and everything appears to be working. I was unable to get the integration tests to run but will try to work on that while this PR is pending review.

I understand this is a massive PR and a huge refactor but I would love to see this get some attention as it was a decent effort. Even being backward compatible it might be worth a 4.x release given potential impact and the number of people that have become dependent on this very popular package.

P.S. Can do normal online PR review, but I actually live and work near DUMBO and could do in-person walkthrough of the changes if that makes sense.

lee-reinhardt commented on index.ts:
```diff
@@ -539,73 +601,85 @@ class ServerlessCustomDomain {
+        let LogicalResourceId = "ApiGatewayRestApi";
+        if (domain.apiType === Globals.apiTypes.http) {
+            LogicalResourceId = "HttpApi";
+        } else if (domain.apiType === Globals.apiTypes.rest) {
```
I missed the new `apiType` config when trying this out and noticed it tried to set up a `WebsocketsApi`. It's caused by comparing to the wrong type here.

```suggestion
        } else if (domain.apiType === Globals.apiTypes.websocket) {
```

TehNrd replied:
Gah, you are absolutely right. I originally had it hardcoded to `"WEBSOCKET"` but just messed up when I flipped it to use the Globals object and typed the wrong thing.

It should be fixed now.

TehNrd replied:
Saw a few other places and fixed them as well, it looks like it was errant find and replace.

TehNrd replied:
Also working on adding some unit tests to catch this.

TehNrd replied:
Unit tests added that would/should have caught this.

lee-reinhardt commented:
Thanks for implementing this! I've tried it out today and can verify it works for the HTTP API type.

Something I've noticed is that subsequent runs of `sls deploy` will fail with this error:
```
Serverless Domain Manager: Error: dev.api.example.com:  ConflictException: ApiMapping key already exists for this domain name
Serverless Domain Manager: Error: dev.api.example.com:  Error: Error: Unable to create basepath mapping.

```

Ideally it would recognize this situation and proceed as normal. I worked around this in my local testing by doing:

```js
        // Make API call
        try {
          yield this.apigatewayV2.createApiMapping(params).promise();
          this.serverless.cli.log("Created basepath mapping.");
        } catch (err) {
          if (err.code === "ConflictException") {
            this.serverless.cli.log(
              "Basepath already exists",
              "Serverless Domain Manager"
            );
            return;
          }

          this.logIfDebug(err, domain.givenDomainName);
          throw new Error(`Error: Unable to create basepath mapping.\n`);
        }
```

I have no idea if that is reasonable for all cases or not, but it worked for my situation.

TehNrd commented:
@lee-reinhardt That issue should be fixed now. Thanks for jumping on this and doing a little testing, really appreciate it!

zainzafar commented:
When can we expect this to be merged?

cbschuld commented:
Curious; does anyone know if we are able to install @TehNrd 's PR directly via NPM.  I installed it with
```
npm install "TehNrd/serverless-domain-manager" --save-dev
```

but I end up with the error:

```
Serverless plugin "serverless-domain-manager" not found. Make sure it's installed and listed in the "plugins" section of your serverless config file.
```

Bummer!

_Also, super excited to see this PR pushed into a new release_

TehNrd commented:
@cbschuld Unfortunately, that is not going to work because Amplify does not keep the compiled distribution files in the repository. They seem to have some sort of separate NPM publishing process not included in this repo.

Here is something you can do for testing, but it isn't really scalable.

- pull down the `TehNrd/serverless-domain-manager` branch to local project
- In the project directory run commands `npm install` then `npm run build`
- In the serverless project that you want to test this in, add the following to the `plugins:` The full path will be different for you, but you can easily get it by opening the project in VSCode or similar, right click on the /dist/index.js file and "Copy Path"
<img width="385" alt="Screen Shot 2020-03-26 at 10 49 13 AM" src="https://user-images.githubusercontent.com/1007085/77660356-99b38500-6f4f-11ea-8ad0-c81b51e26efa.png">
```
plugins:
  - /Users/jason/Documents/Projects/serverless-domain-manager/dist/index.js
```

cbschuld commented:
@TehNrd - awesome; thank you!  This will get my by while I wait for this PR to be eaten.

cbschuld commented:
@TehNrd - I am bumping into a few things - unsure if I am missing something and being stupid or if it is underscoring something else

converted customDomain to:
```
  customDomain:
    domainName: ${self:custom.domains.${self:custom.stage}}
    stage: ${self:custom.stage}
    basePath: ""
    certificateName: "*.DOMAIN.COM"
    createRoute53Record: true
    endpointType: 'regional'
    securityPolicy: tls_1_2
    apiType: http
```

It will not deploy with error:
```
  Error: Error: Failed to find CloudFormation resources for dev-api.DOMAIN.COM

      at ServerlessCustomDomain.<anonymous> (/projects/serverless-domain-manager/dist/index.js:559:23)
      at Generator.throw (<anonymous>)
      at rejected (/projects/serverless-domain-manager/dist/index.js:5:65)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

While in debug mode I am seeing this:
```
Serverless: [AWS cloudformation 400 0.266s 0 retries] describeStackResource({
  LogicalResourceId: 'ApiGatewayRestApi',
  StackName: 'api-DOMAIN-COM-dev'
})
Serverless Domain Manager: ValidationError: Resource ApiGatewayRestApi does not exist for stack api-DOMAIN-COM-dev
```

any help is both appreciated & awesome!  Thx!

TehNrd commented:
@cbschuld Looks like you pulled down and ran build on the master branch which is currently the released version and with no support for http/ws.

If you pull down the whole repo you'll need to switch to the `http-ws-support` branch and then do `npm run build`

Can also pull this branch directly, `https://github.com/TehNrd/serverless-domain-manager/tree/http-ws-support`

I'll update post above to be more clear.

cbschuld commented:
@TehNrd - [ shakes head in shame ] - thank you - that was it; apologies; was in the wrong branch - I'll push ahead

cbschuld commented:
@TehNrd - apologies, I keep bumping into issues ; I was hoping I could be a good test case for the updates and I feel like the whiny dev w/o enough context to help.

Here is the error I am seeing:

```
Serverless: [AWS cloudformation 400 0.256s 0 retries] describeStackResource({ LogicalResourceId: 'HttpApi', StackName: 'api-DOMAIN-COM-dev' })
Serverless Domain Manager: Error: dev-api.DOMAIN-COM:  ValidationError: Resource HttpApi does not exist for stack api-DOMAIN-COM-dev
```

I checked the stack and the resource is definitely not there.

Here is additional context on the error:

```
service: api-DOMAIN-COM
stage: dev
region: us-west-2
stack: api-DOMAIN-COM-dev
resources: 6
api keys:
  None
endpoints:
  None
functions:
  test: api-DOMAIN-COM-dev-test
layers:
  None
Serverless: [AWS cloudformation 400 0.256s 0 retries] describeStackResource({ LogicalResourceId: 'HttpApi', StackName: 'api-DOMAIN-COM-dev' })
Serverless Domain Manager: Error: dev-api.DOMAIN.COM:  ValidationError: Resource HttpApi does not exist for stack api-DOMAIN-COM-dev
Serverless Domain Manager: Error: dev-api.DOMAIN.COM:  Error: Error: Failed to find CloudFormation resources for dev-api.DOMAIN.COM


  Error --------------------------------------------------

  Error: Error: Unable to setup base domain mappings for dev-api.DOMAIN.COM
      at ServerlessCustomDomain.<anonymous> (/private/var/www/serverless-domain-manager/dist/index.js:137:27)
      at Generator.throw (<anonymous>)
      at rejected (/private/var/www/serverless-domain-manager/dist/index.js:5:65)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

again, I am in hopes I am just not thinking crisply around this challenge

TehNrd commented:
No stress on issues, happy to find and them and fix! 

Well, that is a more valid error message as it is on the right branch. 🤔 

Was this for an existing domain that was already existing or a new one?

Did you run `sls create_domain`?

cbschuld commented:
@TehNrd  - yeah; domain is definitely created.  Also, I removed everything and started from a clean slate and ended up w/ the same error.  Here is the call from create_domain for sanity:

```
Serverless: Load command output:get
Serverless: Load command output:list
Serverless: Load command param
Serverless: Load command param:get
Serverless: Load command param:list
Serverless: Load command dev
Serverless: Invoke create_domain
Serverless: [AWS apigateway 200 0.369s 0 retries] getDomainName({ domainName: 'dev-api.DOMAIN.COM' })
Serverless: Custom domain dev-api.DOMAIN.COM already exists.
```

cbschuld commented:
> No stress on issues, happy to find and them and fix!
> 
> Well, that is a more valid error message as it is on the right branch. 🤔
> 
> Was this for an existing domain that was already created or a new one?
> 
> Did you run `sls create_domain`?

Existing domain.

TehNrd commented:
An existing domain might be the issue? How long ago and how was it created?

Older domains will not work as they are created with AWS APIGateway API (v1) that is not compatible with HTTP APIs as the Domains needed to be created with AWS ApiGatewayV2 API. AWS doesn't make this obvious at all as there is no way to see in the console what API version was used to create a domain. Even there APIs don't tell you what version a Domain Name was created with!

If it was created and used with an older REST API, it won't be usable with the new HTTP APIs.

Let me know if this was the issue and maybe I can think of a creative way to solve this.

cbschuld commented:
@TehNrd  - copy that.  It was definitely created with the previous version.  The domain's route53 record is pretty packed though so re-creating will bit a bit of a pain.  I suspect I won't be the only one looking to convert REST subdomains over to HTTP either.  Bummer.  I am contemplating blowing away the domain just for pure testing reasons.

TehNrd commented:
Ya, it's definitely an issue. Let me think about the best way to handle it. My first thought is there probably isn't a good way as this needs to be done in a controlled manner due to all the Route 53 and DNS propagation stuff that needs to happen. It would be pretty bad if a deploy swapped out DNS stuff unexpectedly.

cbschuld commented:
@TehNrd - resetting everything back to zero I am noticing that the v2 version * may * not be dealing with edge vs regional certificates in us-east-1 vs let's say us-west-2 for certificates.  Obviously the provider is `this.serverless.providers.aws.sdk.ApiGatewayV2` Which may end up impeding the creation of the HttpApi downstream.  I am still digging but not finding a root cause (yet).

I have yet to get it to create a domain correctly.

TehNrd commented:
Edge vs Regional certificate? Are there different types of certs? Or do you mean certs stored in two different regions?

FWIW, Domains that use an HTTP API can only be Regional, not Edge. (not sure if that is related)

TehNrd commented:
@cbschuld Should be fixed now, but let me know.

Route 53 is one of my weaker areas of AWS knowledge and the issue was that I/code was not handling the new response object correctly for domains created with ApiGatewatV2 and it was setting the `DomainInfo.domainName` incorrectly. It was falling back to the `givenName` from the `customDomain` serverless.yml file when it should have been using the `ApiGatewayDomainName` property in the response. 

Thanks for testing, catching this, and jumping on Zoom to dig into details!

TehNrd commented:
I started to look into how this could support migration from an older REST API to an HTTP API as an older domain even created with APIGateway (v1) should still be able to create base path mappings to HTTP APIS if it is a Regional Domain.

Investigating this I discovered an issue (easily fixable) with the current implementation. It assumes if setting up a REST API, use APIGateway (V1) and if HTTP or WebSocket API use ApiGatewayV2. Unfortunately, It's not quite this simple as Domains can be of the type Edge Optimized or Regional, but the REST APIs can also be configured Regional/Edge/Private. This creates a little matrix of possibilities of 1) is the combination even valid/allowed, and 2) if the combination is valid what APIGateway API version should be used for configuration, v1 or v2.

It think this table handles all the scenarios but I need to actual testing.

If it is correct, I will add upfront validation to throw an error the `customDomain` config is invalid. If it is valid a different method will determine what ApiGateway API version to use.

|          |               | Domain Type |          |
|----------|---------------|-------------|----------|
|          |               | EDGE        | REGIONAL |
| **Api Type** | REST Edge     |     Y v1    |     N    |
|          | REST Regional |      N      |   Y v2   |
|          | REST Private  |      ?      |     ?    |
|          | WS            |      N      |   Y v2   |
|          | HTTP          |      N      |   Y v2   |

cbschuld commented:
@TehNrd - I may have located a definitive answer on custom domains on private REST (https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-private-apis.html) it shares that "Custom domain names are not supported for private APIs."  Thus, I believe that can simply be dropped from the matrix.  A continued "Nice work" BTW!  I am continuing to work with it here and will continue using your PR & report back whatever I find.

cbschuld commented:
@TehNrd - working a-okay so far; not finding issues right now relative to my use cases.

As an aside - you can have an empty basePath in the customDomain / http area - if you do add one you cannot add any type of slash.

**Valid**
```
  customDomain:
    http:
      domainName: ${self:custom.domains.${self:custom.stage}}
      stage: ${self:custom.stage}
      basePath: "v1"
      certificateName: "domain.com"
      createRoute53Record: true
      endpointType: 'regional'
      securityPolicy: tls_1_2
```

Invalid:
```
  customDomain:
    http:
      domainName: ${self:custom.domains.${self:custom.stage}}
      stage: ${self:custom.stage}
      basePath: "/v1"
      certificateName: "domain.com"
      createRoute53Record: true
      endpointType: 'regional'
      securityPolicy: tls_1_2
```

cbschuld commented:
@TehNrd - also, additionally, not sure we can work around this - if you push a domain with a basePath mapping you cannot update it - you have to remove it and re-create 👎

TehNrd commented:
> @TehNrd - also, additionally, not sure we can work around this - if you push a domain with a basePath mapping you cannot update it - you have to remove it and re-create 👎

Is it throwing an error? Updates should definitely work.

cbschuld commented:
> > @TehNrd - also, additionally, not sure we can work around this - if you push a domain with a basePath mapping you cannot update it - you have to remove it and re-create 👎
> 
> Is it throwing an error? Updates should definitely work.
> 
> Did you mean push _**without**_ a basePath mapping?

@TehNrd  - Yes, updated from a set basePath to an empty basePath:

```
Serverless: [AWS apigatewayv2 409 0.744s 0 retries] updateApiMapping({
  ApiId: 'tg74x4mqac',
  ApiMappingId: 'xu2g2m',
  ApiMappingKey: '(none)',
  DomainName: 'dev-domain.com',
  Stage: '$default'
})
Serverless Domain Manager: Error: dev-domain.com:  ConflictException: Cannot update the ApiMapping key to empty because this domain name already has another ApiMapping. Only one ApiMapping is allowed if the key is empty.
Serverless Domain Manager: Error: dev-domain.com:  Error: Error: Unable to update basepath mapping.
```

ricardovf commented:
Any preview on when it will be merged?

TehNrd commented:
@cbschuld Not sure if we can handle that particular error as it is coming from AWS.

I am looking into removing as much Apigateway V1 references as possible and that may help, might also help from migrating from REST to HTTP, but the AWS documentation is not surer helpful at the moment. It seems like v2 can do everything v1 can do except create EDGE domain names, even though the documentation [here](https://docs.aws.amazon.com/cli/latest/reference/apigatewayv2/create-domain-name.html), [here](https://docs.aws.amazon.com/apigatewayv2/latest/api-reference/domainnames-domainname.html), and [here](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/ApiGatewayV2.html#createDomainName-property) says it should be possible.  I've opened a case with AWS support to get this answered.

Once this last bit is finalized I'll create a new branch with the compiled code that should make test/using this PR easier.

cbschuld commented:
FWIW as a continued update on this PR - I have been using @TehNrd 's PR/branch including all of the recent updates & have had a lot of success with it.  (thx Jason)

TehNrd commented:
@cbschuld thanks! I actually just pushed one, hopefully last, larger update and I'd recommend using that. The logic before was to segment domain management based on API type but this was incorrect. It should have segmented functionality on Domain type (regional vs edge) and it does this now.

These changes also make it possible to use an existing domain and change it from a REST to HTTP API. In the readme, there is a new section called "Changing API Types" that explains how this can be done.

For anyone else that wants to test I also created a build that is can be installed with npm:
```json
"dependencies": {
  "serverless-domain-manager": "tehnrd/serverless-domain-manager#http-ws-support-BETA"
},
```

cbschuld commented:
> For anyone that wants to test this PR I also created a build that is can be installed with npm:

I updated to this BETA dependency, rebuilt, re-tested and re-released - no issues in my project _</anecdote>_

chandu1310 commented:
Is there an ETA for this PR merge ?

Sparticuz commented:
Trying to migrate one of my projects from REST to HTTP. I found it much easier to go here, https://console.aws.amazon.com/apigateway/main/publish/domain-names to change the endpoint type to regional. However, I'm still having trouble deploying.
```
Serverless: [AWS sts 200 0.244s 0 retries] getCallerIdentity({})
Serverless: [AWS apigateway 200 0.369s 0 retries] getRestApis({ position: undefined, limit: 500 })
Serverless: WARNING: "allowPathMatching" is set for dev-api.******.com.
                    This should only be used when migrating a path to a different API type. e.g. REST to HTTP.
Serverless: [AWS cloudformation 200 0.212s 0 retries] describeStackResource({ LogicalResourceId: 'HttpApi', StackName: '******-api-dev' })
Serverless: [AWS apigatewayv2 200 0.234s 0 retries] getApiMappings({ DomainName: 'dev-api.******.com' })
Serverless: [AWS apigatewayv2 400 0.226s 0 retries] createApiMapping({
  ApiId: '***********',
  ApiMappingKey: '(none)',
  DomainName: 'dev-api.******.com',
  Stage: '$default'
})
Serverless Domain Manager: Error: dev-api.******.com:  BadRequestException: Only REGIONAL domain names with a security policy of TLS 1.2 can be managed through the API Gateway V2 API. For TLS 1.0 domain names, please use the API Gateway V1 API to upgrade your security policy to TLS 1.2. Currently, only REST APIs can be attached to a domain name with a security policy of TLS 1.0.
Serverless Domain Manager: Error: dev-api.******.com:  Error: Error: dev-api.******.com: Unable to create basepath mapping.
```
What's odd is that it shows both endpoint types on the api gateway custom domain name page. Is this why it's not deploying? Have I just not waited long enough for the endpoint type change? I noticed in the dev guide it said it takes a minute, but I've waited longer than that.
![image](https://user-images.githubusercontent.com/161913/78674133-272c9880-78b1-11ea-8736-72d769f1e314.png)

TehNrd commented:
@Sparticuz That is expected and an error from AWS. The Regional endpoint type can be added in the UI, but see how it is still showing the Edge endpoint type? You need to remove that from the domain name first. I'm not sure this can be done through the Console. You could try doing a quick edit/save to see if it drops, but it might need to be done through the AWS API.

Sparticuz commented:
@TehNrd OK, in the console I did a delete on that custom domain name, then `sls create_domain` and it seems to have created it properly, although it did change the execute-api url, which doesn't really matter since I'm using a custom domain name 😄

TehNrd commented:
@Sparticuz Deleting and re-creating the domain name is one way, but that would not be ideal in a production environment without the potential for serious downtime due to DNS propagation.

pimlie commented:
@TehNrd Thanks for this PR.

Have been trying out your beta but unfortunately I get a TypeError when trying to remove my stack:
```
Serverless Domain Manager: Error: www.example.com:  TypeError: Cannot read property 'ApiMappingId' of undefined

  Error --------------------------------------------------

  Error: Error: Unable to remove base bath mappings for domain www.example.com
      at ServerlessCustomDomain.<anonymous> (/var/project/node_modules/serverless-domain-manager/dist/index.js:153:27)
      at Generator.throw (<anonymous>)
      at rejected (/var/project/node_modules/serverless-domain-manager/dist/index.js:5:65)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
      at runNextTicks (internal/process/task_queues.js:62:3)
      at processImmediate (internal/timers.js:412:9)

```
when I log the domain in _deleteBasePathMapping_ it prints:
```js
DomainConfig {
  enabled: true,
  givenDomainName: 'www.example.com',
  hostedZonePrivate: undefined,
  certificateArn: undefined,
  certificateName: undefined,
  createRoute53Record: true,
  hostedZoneId: undefined,
  allowPathMatching: undefined,
  basePath: '(none)',
  stage: 'development',
  endpointType: 'REGIONAL',
  apiType: 'HTTP',
  securityPolicy: 'TLS_1_2'
}
```

TehNrd commented:
@pimlie Can you share your serverless.yml file?

kleanthist commented:
I received exactly the same error, using the BETA branch. 
Is this the correct structure:

`domainName: "driver.*********.com"
 stage: ${self:provider.stage}
 basePath: ${self:provider.stage}
 certificateName: "*.*********.com"
 createRoute53Record: true
 endpointType: 'regional'
 securityPolicy: tls_1_2`

Is still required to use http, websockets or rest at the top of ?

pimlie commented:
This is a stripped down version of my config:
```
service: ssr

provider:
  name: aws

functions:
  srr:
    handler: ssr.handler
    name: srr
    package:
      artifact: dist/ssr.zip
      individually: true
    events:
      - httpApi: '*'
  api:
    handler: api.handler
    name: api
    package:
      artifact: dist/api.zip
      individually: true
    events:
      - httpApi:
          method: POST
          path: /api

plugins:
  - serverless-certificate-creator
  - serverless-domain-manager

custom:
  customCertificate:
    certificateName: www.example.com
    rewriteRecords: true
  customDomain:
    domainName: www.example.com
    basePath: ''
    endpointType: regional
    securityPolicy: tls_1_2
    createRoute53Record: true
    apiType: http
```

TehNrd commented:
@pimlie @kleanthist This should be fixed now, on the PR and the installable beta. 

Run `npm update` or `npm install` and it should pull down that latest. Can use `npm list` to make sure it shows beta.1.

```
@ /Users/jason/Documents/Projects/sls-everything
└─┬ serverless-domain-manager@3.4.0-beta.1 (github:tehnrd/serverless-domain-manager#ac27340448c076daa0f8b30f56a44195a9276cb5)```

pimlie commented:
@TehNrd It seems the `Error: Unable to remove base bath mappings for domain` is still showing when you call `sls remove` twice.

While creating/debugging my config I sometimes have errors that prevents `sls remove` from completing, but after the first run `sls remove` will always fail due to this errors:
```
Serverless: [AWS apigatewayv2 404 0.805s 0 retries] getApiMappings({ DomainName: 'www.example.com' })
Serverless Domain Manager: Error: www.example.com:  NotFoundException: Invalid domain name identifier specified
Serverless Domain Manager: Error: www.example.com:  Error: Error: Unable to get API Mappings for www.example.com

  Error --------------------------------------------------
 
  Error: Error: Unable to remove base bath mappings for domain www.example.com
  ...
```

TehNrd commented:
@pimlie Should be fixed now, but there is a good chance the serverless framework itself is going to throw an error. A partially deleted Cloudformation Stack may require some manual intervention to resolve the issue.

reganjohnson commented:
Thanks for the update - this is working for my use case on httpApi. Hope this gets merged in soon.

TehNrd commented:
I'm starting to think this project might be stagnating without any primary maintainers. Is anyone from Amplify looking at this?

https://github.com/amplify-education/serverless-domain-manager/graphs/contributors

There hasn't been any significant activity in months and it seems like @captainsidd, who was the primary maintainer is no longer with Amplify.

If anyone from Amplify is listening, I would definitely be willing to take ownership of the project. I have become intimately familiar with the codebase working on this PR. Being a maintainer here or transferring the project both work for me.

cbschuld commented:
@TehNrd - I am in hopes that @jconstance-amplify maybe able to provide some feedback here as well...  I am in hopes they turn it over to you because I'd love to see this PR merged in.

breath103 commented:
Actually i've been just using tehnrd/serverless-domain-manager#http-ws-support-BETA this on production :) please don't break it!

reganjohnson commented:
Best if you fork it and use your own version - update as needed. I am also
using the beta successfully.

On Tue, Apr 21, 2020 at 12:00 AM Kurt <notifications@github.com> wrote:

> Actually i've been just using
> tehnrd/serverless-domain-manager#http-ws-support-BETA this on production :)
> please don't break it!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/amplify-education/serverless-domain-manager/pull/319#issuecomment-616937275>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC3WU6GHNNOJH2KMBRK7GCDRNUK43ANCNFSM4LDTDZ6Q>
> .
>

TehNrd commented:
> Actually i've been just using tehnrd/serverless-domain-manager#http-ws-support-BETA this on production :) please don't break it!

Ha! Glad to hear it's working and I don't plan on removing it. But, ya, you can just fork my repo and then point your package.json to your repo.

TehNrd commented:
Looks like the Amplify team is starting to merge old PRs...which is good...but it's introduced some conflicts and it's going to take some time for me to merge/resolve all of those back into this, kinda hoping it would have been in the opposite order.

TehNrd commented:
Rebased with the latest commits from master and resolved conflicts.

Also added more CloudFormation outputs as there could be more if you deploy HTTP, WS, and REST all at once. Kept the output names the same for REST for backward compatibility.

SebastianEdwards commented:
@TehNrd just wanted to say I've been using your repo with no issues - thanks for your work on this!

ricardovf commented:
Please merge this!

aj47 commented:
Need this!!

cbschuld commented:
@jconstance-amplify @aoskotsky-amplify this PR has been really solid for me as well.  If it can be merged that would be great.  Sorry to ping you directly but the project feels like it maybe stagnating so I thought I'd ping directly (or maybe this is normal cadence and I am antsy)?

TehNrd commented:
I would love to see this get merged but I also hope to see some planning forethought with the order in which PRs are merged. This PR is a big change and a bit of an architectural change. Trying to rebase this against other smaller and easier to merge PRs may be a challenge.

For example, I am concerned about #327 #88 and how that would impact this PR, perhaps in a very incompatible way requiring a lot of work. 

I think the ideal scenario would be to get this PR merged and then other PRs would need to rebase and make the changes compatible with this.

TehNrd commented:
...and actually I think this PR is just a small tweak away from supporting multiple domains as it already splits the different HTTP types into an array and runs them through a loop. It just needs to have the option to define multiple domains for a given API type and then it could just run the logic again for each additional domain.

aoskotsky-amplify commented:
Hi All,
We are looking at this PR. Sorry for the long delay. We'll hold off merging other PRs so this one can get in first. @TehNrd are there any other changes you are adding to this or is this good to merge?

aoskotsky-amplify commented:
Specifically I mean, do you want to merge this first or get it working with #327 first?

TehNrd commented:
> do you want to merge this first or get

Not going to make any changes, I decided to hold off.

I or @ConradKurth can do different PR to address #327. While I think it would be easy to add, it is technically a different feature, so a separate clean commit would be good in case a revert is needed. And would also make viewing the changes to support multiple domains easier to see rather than burying them in this.

ConradKurth commented:
@TehNrd I agree, lets get this in and then we address the other PR

aoskotsky-amplify commented on index.ts:
```diff
@@ -316,68 +362,87 @@ class ServerlessCustomDomain {
+
+            // Make API call to create domain
+            try {
+                // If creating REST api use v1 of api gateway, else use v2 for HTTP and Websocket
```
Is this comment still accurate? This seems to decide which API to use based on endpoint type and not API type

aoskotsky-amplify replied:
Also, why can't EDGE domains/rest APIs be created using the v2 API?

claygorman replied:
seems like it can https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/ApiGatewayV2.html#createDomainName-property

TehNrd replied:
@aoskotsky-amplify @claygorman That inline comment is incorrect and should say if creating an EDGE API (only REST can be EDGE, so it's a little confusing, but if REST is regional it can us v2 API), but.....

...the AWS docs are wrong, and not yet updated. I opened a case with AWS Support about this issue exactly. 

v1 API cannot create regional domain names.

https://github.com/amplify-education/serverless-domain-manager/pull/319#issuecomment-606131803

<img width="1023" alt="Screen Shot 2020-05-01 at 7 39 08 PM" src="https://user-images.githubusercontent.com/1007085/80849077-a0cf5200-8be3-11ea-91d5-4306b18b91dd.png">
<img width="995" alt="Screen Shot 2020-05-01 at 7 39 17 PM" src="https://user-images.githubusercontent.com/1007085/80849079-a3ca4280-8be3-11ea-961a-835304fbb0f5.png">
<img width="989" alt="Screen Shot 2020-05-01 at 7 39 23 PM" src="https://user-images.githubusercontent.com/1007085/80849083-a6c53300-8be3-11ea-8f93-ac965b8da4ff.png">
<img width="988" alt="Screen Shot 2020-05-01 at 7 39 40 PM" src="https://user-images.githubusercontent.com/1007085/80849089-aa58ba00-8be3-11ea-8068-a250c52a79ac.png">

TehNrd replied:
I think we are good to resolve this one? @aoskotsky-amplify

TehNrd commented:
Thanks @aoskotsky-amplify, I'll give those a look this evening.

Do you have a preference for the process of implementing/fixing feedback?
1) Make changes and force push to update
or
2) Make a new commit to easily see the changes, and then squash those into one commit before merge.

(embarrassed about all those typos and just installed VSCode spellchecker!)

aoskotsky-amplify commented:
@TehNrd I think option 2 makes sense. Just send a new commit with the changes. Thanks again!

TehNrd commented on index.ts:
```diff
@@ -75,199 +65,255 @@ class ServerlessCustomDomain {
+                }
+
+                await this.getDomainInfo();
+                // this.addOutputs(domain);
```
Yup, an artifact of merging in the #320 changes. 

Removed.

TehNrd commented on index.ts:
```diff
@@ -316,68 +362,87 @@ class ServerlessCustomDomain {
+    public async getDomainInfo(): Promise<void> {
+        await Promise.all(this.domains.map(async (domain) => {
+            try {
+                const domainInfo = await this.apigatewayV2.getDomainName({
```
In this case, the v2 API works and I figured it best to use it in as many places as possible in an attempt to standardize the responses and the code that uses these responses. 

The v2 API can read existing Domain and API Mappings without issue. It's only the creation of REST EDGE domains and API Mappings that it can't handle.

TehNrd commented on DomainConfig.ts:
```diff
@@ -0,0 +1,100 @@
+     * If the property's value is provided, this should be boolean, otherwise an exception is thrown.
+     * If no customDomain object exists, an exception is thrown.
+     */
+    public evaluateEnabled(enabled: any): boolean {
```
Yup, changed to private.

TehNrd commented on index.ts:
```diff
@@ -630,26 +768,23 @@ class ServerlessCustomDomain {
      * @param message message to be printed
      */
-    public logIfDebug(message: any): void {
+    public logIfDebug(message: any, domain?: string): void {
```
It can, totally up to the Amplify team. I can rename it but I will wait for confirmation from you.

This was an already existing method I just added the domain argument to so the user would know which domain was failing on a multi API deployment.

aoskotsky-amplify replied:
I was referring to the fact that the message now always includes the word `Error` so we could make the function name show that. It's a nitpick so leaving it as is for now is fine too

TehNrd replied:
Ah, gotcha. I checked `master` it is only using for errors, but I'm inclined to keep it as is for now as it could theoretically be used for other messages. 

And there isn't a dedicated `serverless.cli` method specifically for errors:

<img width="511" alt="Screen Shot 2020-05-04 at 11 13 18 PM" src="https://user-images.githubusercontent.com/1007085/81032219-0c702400-8e5d-11ea-9505-a78f5976be30.png">

...so probably best if this is handled in its own separate commit.

TehNrd commented:
@aoskotsky-amplify I fixed and resolved the simple changes.

I left open the others that are a little more complex or need more discussion and will let Amplify mark as Resolved when appropriate.

aoskotsky-amplify commented on index.ts:
```diff
@@ -75,199 +65,254 @@ class ServerlessCustomDomain {
+                }
+            } catch (err) {
+                this.logIfDebug(err, domain.givenDomainName);
+                throw new Error(`Error: Unable to craete domain ${domain.givenDomainName}`);
```
```suggestion
                throw new Error(`Error: Unable to create domain ${domain.givenDomainName}`);
```

aoskotsky-amplify commented on index.ts:
```diff
@@ -75,199 +65,254 @@ class ServerlessCustomDomain {
+                        API Mappings may need to be manually removed.`, "Serverless Domain Manager");
+                } else {
+                    this.logIfDebug(err, domain.givenDomainName);
+                    throw new Error(`Error: Unable to remove base bath mappings for domain ${domain.givenDomainName}`);
```
This is actually changing some of the old behavior. This used to log an error without throwing an exception. I ran into this by accident by running `delete_domain` and `remove` out of order. For example:

```
sls create_domain
sls deploy
sls delete_domain
sls remove
```

The `sls remove` command above started throwing errors and I wasn't able to remove the API until running `sls create_domain` and `sls_deploy` again. 

I see the desire to be more strict with the error handling so can either change this back to a log statement or add a note to the readme.

TehNrd replied:
Hmm...ya...I'm looking at that and can't recall a good reason why. I think I was just following the same approach used in other try/catch blocks and may have been a copy/paste error.

I removed the `throw` and added a log.

aoskotsky-amplify commented on index.ts:
```diff
@@ -75,199 +65,254 @@ class ServerlessCustomDomain {
+        this.route53 = new this.serverless.providers.aws.sdk.Route53(credentials);
+        this.cloudformation = new this.serverless.providers.aws.sdk.CloudFormation(credentials);
+
+        // Loop over the domain configurations and popluates the domains array with DomainConfigs
```
```suggestion
        // Loop over the domain configurations and populate the domains array with DomainConfigs
```

TehNrd commented:
@aoskotsky-amplify I think I addressed the issues found. Let me know when you want this squashed down to one commit before any merging (or other issues to fix)

TehNrd commented:
No prob, happy to help!

You beat me to it! I was going to smash those commits into one for a little cleaner history, but not a big deal. You can do it on master if you want before anyone else starts working off of it.

aoskotsky-amplify commented:
I think it's fine having the longer history. Thanks. Working on updating the package version and change log now.

reganjohnson commented:
Hey @TehNrd hoping you can help me here. Getting this error when deploying - have tried to delete the domain, mappings, etc, and recreate using SLS.

`Serverless: [AWS cloudformation 400 0.168s 0 retries] describeStackResource({ LogicalResourceId: 'HttpApi', StackName: 'api' })
Serverless Domain Manager: Error: api.domain.com:  ValidationError: Resource HttpApi does not exist for stack api
Serverless Domain Manager: Error: api.domain.com:  Error: Error: Failed to find CloudFormation resources for api.domain.com


  Error --------------------------------------------------

  Error: Error: Unable to setup base domain mappings for api.domain.com
      at ServerlessCustomDomain.<anonymous> (node_modules/serverless-domain-manager/dist/index.js:142:27)
      at Generator.throw (<anonymous>)
      at rejected (node_modules/serverless-domain-manager/dist/index.js:5:65)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)`

Any advice where to go next?

TehNrd commented:
@reganjohnson You using 4.x? https://www.npmjs.com/package/serverless-domain-manager

What does your config look like?

reganjohnson commented:
@TehNrd Yes to 4.0

```
service: api

provider:
  name: aws
  region: us-east-1
  stage: dev
  runtime: provided
  tags:
    Environment: dev
  httpApi:
    cors: true
  logs:
    httpApi: true
  vpc:
    securityGroupIds:
      - sg-f123456
    subnetIds:
      - subnet-01111111111
      - subnet-01111111112
      - subnet-01111111113
  iamRoleStatements:
    - Effect: Allow
      Action:
        - sqs:SendMessage
        - sqs:ReceiveMessage
        - sqs:GetQueueAttributes
      Resource: arn:aws:sqs:us-east-1:123456789:sqs-dev
  environment:
    ENV_VAR_1: value

plugins:
  - ./vendor/bref/bref
  - serverless-domain-manager

custom:
  customDomain:
    domainName: api.domain.com
    basePath: ''
    stage: dev
    apiType: http
    createRoute53Record: false
    endpointType: regional
    securityPolicy: tls_1_2
    certificateArn: arn:aws:acm:us-east-1:123456789:certificate/uuid

package:
  exclude:
    - node_modules/**
    - etc...

functions:
  api:
    handler: public/index.php
    timeout: 10
    memorySize: 1024
    layers:
      - ${bref:layer.php-74-fpm}
    events:
      - httpApi: '*'

  queue-listener:
    handler: public/worker.php
    timeout: 120
    reservedConcurrency: 10
    events:
      - sqs: arn:aws:sqs:us-east-1:123456789:sqs-dev
        batchSize: 1
    layers:
      - ${bref:layer.php-74}
```

TehNrd commented:
Try dropping the `stage` option. It technically isn't required for HTTP apis. It should inherit the stage used with `sls deploy`

If there is specific need for it, let me know as I'm curious.

Related to https://github.com/amplify-education/serverless-domain-manager/issues/334

reganjohnson commented:
> Try dropping the `stage` option. It technically isn't required for HTTP apis. It should inherit the stage used with `sls deploy`
> 
> If there is specific need for it, let me know as I'm curious.
> 
> Related to #334

No need for the stage on the new httpApi - was using it for the REST setup I migrated from.

This appears to have worked! Thank you much for the quick reply.

yasso1am commented:
Hi there, it appears this PR accomplished exactly what I need, however in practice I am unable to receive a Cloudfront distribution link through the process of running `sls create_domain` and `sls deploy`.

I am attempting to deploy an Api Gateway V2, as in an http api vs rest api. From my reading, it appears I can only do this with regional endpoint types, since edge-optimized are not supported. I created my certificate in the correct region, and validated successfully through DNS validation. I am NOT using Route53, but instead using Cloudflare. As such, I have set `createRoute53Record` to `false`.

Here is the relevant part of my serverless.yml file.

```
custom:
  customDomain:
    domainName: dev.domain.com
    basePath: api
    apiType: http
    createRoute53Record: false
    endpointType: regional
    autoDomain: true
    securityPolicy: tls_1_2
```

When I run `sls create_domain`, it does create a custom domain in my api gateway, however there is no console output. All of the documentation/tutorials I follow says I should receive something. When I run this command with the `--verbose` flag I receive the following:

```
sls create_domain --verbose
Running "serverless" from node_modules
'dev.domain.com' does not exist.
Searching for a certificate with the 'dev.domain.com' domain
Custom domain 'dev.domain.com' was created.
                 New domains may take up to 40 minutes to be initialized.
Creating/updating route53 record for 'dev.domain.com'.
Skipping creation of Route53 record.
```
 
 So I believe this has worked successfully, however when I then run `sls deploy`, the output in the `Serverless Domain Manager:` section is not providing me a Target Domain of a Cloudfront distribution, but instead of the api gateway custom domain. It looks like this:
 
 ```
 Serverless Domain Manager:
  Domain Name: dev.domain.com
  Target Domain: d-randomId.execute-api.us-west-2.amazonaws.com
  Hosted Zone Id: Z2OJLYMUO9EFXC
```

It is my understanding that I should be receiving a Cloudfront distribution url in the Target Domain section of this output, which I can use to configure my DNS in Cloudflare as an Alias/CNAME to point to it. Why is this not happening? It appears to me that @reganjohnson is not creating a route53 record either, so I don't think this should be my issue, what gives?

adminy commented:
same here, target domain for rest seems to work fine, but for http api target domain is not the same as endpoints listed after sls deploy.
